src/garage/envs/garage_env.py

```diff
@@ -80,9 +87,16 @@ def __new__(cls, *args, **kwargs):
 
         return super(GarageEnv, cls).__new__(cls)
 
-    def __init__(self, env=None, env_name='', is_image=False):
+    def __init__(self,
+                 env=None,
+                 env_name='',
+                 is_image=False,
+                 max_episode_length=math.inf):
```

Comment:
if this environment is wrapped in a `TimeLimit` or its spec has a `max_episode_steps` you should use that instead https://github.com/openai/gym/blob/master/gym/envs/registration.py#L42 (probably a ValueError if you have a collision?)

Reply:
Gym environments created by `gym.make()` must have default `max_episode_steps` in their specs. User can specify this via `gym.make(max_episode_steps=num)`. This is somewhat less intuitive, as `max_episode_length` in `GymEnv` is supposed to do the same thing. 

My suggestion: Prefer `max_episode_length` passed to `GymEnv` over arguments passed to `gym.make()`. If `max_episode_length` is not present, use spec's value instead.
```       
if isinstance(self._env, TimeLimit):  # env is wrapped by TimeLimit
    if math.isinf(max_episode_length):  # use default value
        max_episode_length = self._env._max_episode_steps
    else:  # overwrite gym TimeLimit's max episode steps
        self._env._max_episode_steps = max_episode_length
```

Reply:
> ```py
> if isinstance(self._env, TimeLimit):  # env is wrapped by TimeLimit
> ```
An environment can be wrapped an arbitrary number of levels (and often is), so this doesn't detect all cases. Is there an instance where an environment is wrapped in `TimeLimit` but its `env.spec` doesn't contain the that limit? Perhaps it's easier to just interrogate `env.spec`?

> ```py
> if math.isinf(max_episode_length):  # use default value
> ```
This would be more idiomatic if  the default was `None` and you tested for that instead

> My suggestion: Prefer max_episode_length passed to GymEnv over arguments passed to gym.make(). If max_episode_length is not present, use spec's value instead.

LGTM

Reply:
Inspecting `gym.make()`'s logic here https://github.com/openai/gym/blob/master/gym/envs/registration.py#L95, I don't think `env.spec` will not contain a limit. 

> This would be more idiomatic if the default was None and you tested for that instead
Sounds good!

Reply:
> An environment can be wrapped an arbitrary number of levels (and often is), so this doesn't detect all cases. 

In garage, gym environments created by `gym.make()` will be wrapped only by `GymEnv`. Other wrappers in garage are now changed to adapt the `Environment` API. For example, should use `normalize(GymEnv(gym.make()))` rather than `GymEnv(normalize(gym.make()))` So this logic should capture all cases for garage-related gym envs?

Reply:
i just meant that `TimeLimit` is not necessarily the outermost gym wrapper applied (by the user or gym) so can just use `isinstance` to detect the presence of something like `env = FooWrapper(BarWrapper(BazWapper(TimeLimit(CartPole()))))`

As for `max_episode_steps`, did you not find that this property is populated? https://github.com/openai/gym/blob/master/gym/envs/registration.py#L42